docs/getting-started/deployment-docker.md

```diff
@@ -0,0 +1,241 @@
+to be executed with the root of the repo as the build context, in order to get
+access to the root `yarn.lock` and `package.json`, along with any other files
+that might be needed, such as `.npmrc`.
+
+In order to speed up the build we can significantly reduce the build context
+size using the following `.dockerignore` in the root of the repo:
+
+```text
+.git
+node_modules
+packages
+!packages/backend/dist
+plugins
+```
+
+With the project build and the `.dockerignore` and `Dockerfile` in place, we are
+now ready to build the final image. Assuming we're at the root of the repo, we
+execute the build like this:
+
+```bash
+docker image build . -f packages/backend/Dockerfile --tag backstage
+```
+
+To try out the image locally you can run the following:
+
+```sh
+docker run -it -p 7000:7000 backstage
+```
+
+You should then start to get logs in your terminal, and then you can open your
+browser at `http://localhost:7000`
+
+[`backend:bundle` command docs](../cli/commands#backendbundle)
+
+### Multistage Build
+
+This section describes how to set up a multi-stage Docker build that builds the
+entire project within Docker. This is typically slower than a host build, but is
+sometimes desired because Docker in Docker is not available in the build
+environment or other requirements.
+
+The build is split into three different stages, where the first stage finds all
+of the `package.json`s that are relevant for the initial install step enabling
+us to cache the initial `yarn install` that installs all dependencies. The
+second stage executes the build itself, and is similar to the steps we execute
+on the host in the host build. The third and final stage then packages it all
+together into the final image, and is similar to the `Dockerfile` of the host
+build.
+
+The following `Dockerfile` executes the multi-stage build and should be added to
+the repo root:
+
+```Dockerfile
+# Stage 1 - Create yarn install skeleton layer
+FROM node:14-buster-slim AS packages
+
+WORKDIR /app
+COPY package.json yarn.lock ./
+
+COPY packages packages
+COPY plugins plugins
+
+RUN find packages \! -name "package.json" -mindepth 2 -maxdepth 2 -print | xargs rm -rf
+
+# Stage 2 - Install dependencies and build packages
+FROM node:14-buster-slim AS build
+
+WORKDIR /app
+COPY --from=packages /app .
+
+RUN yarn install --frozen-lockfile --network-timeout 600000 && rm -rf "$(yarn cache dir)"D
+
+COPY . .
+
+RUN yarn tsc
+RUN yarn --cwd packages/backend backstage-cli backend:bundle --build-dependencies
+
+# Stage 3 - Build the actual backend image and install production dependencies
+FROM node:14-buster-slim
+
+WORKDIR /app
+
+# Copy the install dependencies from the build stage and context
+COPY --from=build /app/yarn.lock /app/package.json /app/packages/backend/dist/skeleton.tar.gz ./
+RUN tar xzf skeleton.tar.gz && rm skeleton.tar.gz
+
+RUN yarn install --frozen-lockfile --production --network-timeout 600000 && rm -rf "$(yarn cache dir)"
```

Comment:
The previous image deleted the cache too, not much overlap to be gained by keeping it around?

Reply:
Not sure what you mean?

Reply:
Hm that was probably a half baked thought.

I essentially was wondering if the yarn cache from the first image could be leveraged in the second. But that would require explicitly copying it from `build` to here, and paying the cost for that image being a lot larger due to the cache being kept in the fs. Well never mind

Reply:
Yeah it's kind of almost possible, if you could just `COPY && RUN` :p

Reply:
This is possible with the new cache options from buildkit. We just started to use it to cache the yarn cache folder which works great (only locally; we don't cache the cache yet in the CI). We also use it to inject the npm secrets so they don't appear in the image. But I'm not sure if we want to mention it here or even make it the recommendation.

See https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/syntax.md